src/ucs/memory/rcache.c

```diff
@@ -1069,18 +1069,23 @@ static ucs_status_t ucs_rcache_global_list_add(ucs_rcache_t *rcache)
     return status;
 }
 
-static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache) {
+static void ucs_rcache_global_list_remove(ucs_rcache_t *rcache)
+{
+    ucs_async_pipe_t pipe;
+
     pthread_mutex_lock(&ucs_rcache_global_context.lock);
+    pipe = ucs_rcache_global_context.pipe;
     ucs_list_del(&rcache->list);
     if (!ucs_list_is_empty(&ucs_rcache_global_context.list)) {
         pthread_mutex_unlock(&ucs_rcache_global_context.lock);
         return;
     }
 
-    ucs_async_remove_handler(ucs_async_pipe_rfd(
-                                     &ucs_rcache_global_context.pipe),
+    pthread_mutex_unlock(&ucs_rcache_global_context.lock);
+    ucs_async_remove_handler(pipe.read_fd,
                              1);
-    ucs_async_pipe_destroy(&ucs_rcache_global_context.pipe);
+    pthread_mutex_lock(&ucs_rcache_global_context.lock);
```

Comment:
Maybe this lock can be deleted (if no place checks pipe != -1).
If there is such place, pipe should be set to -1 above, after transferring ownership from global to local scope.